cmd/fil-daggregator/main.go

```diff
@@ -0,0 +1,294 @@
+		}
+	}
+
+	if !opts.SkipDagStat {
+		if err := statSources(ctx, opts, toAgg); err != nil {
+			log.Fatal(err)
+		}
+	}
+
+	ramBs := new(rambs.RamBs)
+	ramDs := merkledag.NewDAGService(blockservice.New(ramBs, exchangeoffline.Exchange(ramBs)))
+	root, entries, err := dagaggregator.Aggregate(ctx, ramDs, toAgg)
+	if err != nil {
+		log.Fatalf("aggregation failed: %s", err)
+	}
+
+	if err := writeoutBlocks(ctx, opts, ramBs); err != nil {
+		log.Fatalf("writing newly created dag to IPFS API failed: %s", err)
+	}
+
+	log.Infow("aggregation finished", "aggregateRoot", root, "totalEntries", len(entries))
+}
+
+func statSources(externalCtx context.Context, opts *opts, toAgg []dagaggregator.AggregateDagEntry) error {
+
+	type dagStat struct {
+		Size      uint64
+		NumBlocks uint64
+	}
+
+	ctx, shutdownWorkers := context.WithCancel(externalCtx)
+	defer shutdownWorkers()
+
+	maxWorkers := opts.IpfsAPIMaxWorkers
+	workCh := make(chan int)                // channel of toAgg indexes to work on
+	errCh := make(chan error, 1+maxWorkers) // our own error plus one from each worker
+
+	// work dispenser, watchdog to bail early if an error appears
+	go func() {
+		defer close(workCh)
+
+		toAggIdx := 0
+		for toAggIdx < len(toAgg) {
+			select {
+			case workCh <- toAggIdx:
+				toAggIdx++
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			case err := <-errCh:
+				errCh <- err // put error back where we got it from
+				if err != nil {
+					shutdownWorkers()
+				}
+				return
+			}
+		}
+	}()
+
+	var wg sync.WaitGroup
+
+	for maxWorkers > 0 {
+		maxWorkers--
+		wg.Add(1)
+		go func() {
+			defer wg.Done()
+			api := ipfsapi.NewShell(opts.IpfsAPI)
+			api.SetTimeout(time.Second * time.Duration(opts.IpfsAPITimeoutSecs))
+
+			for {
+				toAggIdx, chanOpen := <-workCh
+				if !chanOpen {
+					return
+				}
+
+				ds := new(dagStat)
+				err := api.Request("dag/stat").Arguments(toAgg[toAggIdx].RootCid.String()).Option("progress", "false").Exec(ctx, ds)
+				if err != nil {
+					errCh <- err
+					return
+				}
+
+				toAgg[toAggIdx].UniqueBlockCount = ds.NumBlocks
+				toAgg[toAggIdx].UniqueBlockCumulativeSize = ds.Size
+			}
+		}()
+	}
+
+	wg.Wait()
+	shutdownWorkers()
+
+	errCh <- externalCtx.Err()
+	return <-errCh
+}
+
+// pulls cids from an AllKeysChan and sends them concurrently via multiple workers to an API
+func writeoutBlocks(externalCtx context.Context, opts *opts, bs blockstore.Blockstore) error {
+
+	ctx, shutdownWorkers := context.WithCancel(externalCtx)
+	defer shutdownWorkers()
+
+	akc, err := bs.AllKeysChan(ctx)
+	if err != nil {
+		return err
+	}
+
+	maxWorkers := opts.IpfsAPIMaxWorkers
+	errCh := make(chan error, 1+maxWorkers) // our own error plus one from each worker
+
+	// watchdog to bail early if an error appears
+	go func() {
+		for {
+			select {
+			case <-ctx.Done():
+				errCh <- ctx.Err()
+				return
+			case err := <-errCh:
+				errCh <- err // put error back where we got it from
```

Comment:
@mvdan especially this is something I am quite unsure of...

Reply:
yeah this smells quite a bit :) I think this would be simplified if you join the "watchdog" goroutine with the main/parent goroutine.